### DISCUSSION TITLE
Build Alpine image via Docker

### DISCUSSION THREAD

**Discussion by owenthereal** [copy/v86#989] [Q&A] [answered]
I wonder whether anyone knows how to build an Alpine image via Docker similar to [tools/docker/debian](https://github.com/copy/v86/tree/master/tools/docker/debian). I'm stuck at mounting the 9p filesystem

**Comment by SuperMaxusa** [comment, marked as answer]
Here is my proof-of-work Alpine Linux on 9pfs, feel free for suggestions.

![alpine](https://github.com/copy/v86/assets/41739128/be6d6c6a-9e8a-4a7a-937f-e09cb5ebe922)
---------------------------------------

**1.** Create Dockerfile
```dockerfile
FROM i386/alpine:3.18.6

ENV KERNEL=lts
ENV HOSTNAME=localhost
ENV ROOT_PASSWORD=root

# Installing base packages, you can add additional packages here (community repo is enabled)
RUN apk add openrc alpine-base agetty alpine-conf

# Install newer mkinitfs from edge (todo: remove this when 3.19+ has worked properly with 9pfs)
RUN apk add mkinitfs --no-cache --allow-untrusted --repository https://dl-cdn.alpinelinux.org/alpine/edge/main/ 

# Installing kernel and remove unused drivers for linux-lts
RUN if [ "$KERNEL" == "lts" ]; then \
    apk add linux-lts linux-firmware-none; \
else \
    apk add linux-$KERNEL; \
fi

# Setting root password and hostname
RUN echo "root:$ROOT_PASSWORD" | chpasswd
RUN setup-hostname $HOSTNAME

# Enable autologin into root on tty1, also you can use autologin.c instead agetty
# (see https://wiki.alpinelinux.org/wiki/TTY_Autologin#By_compiling_your_own_autologin_wrapper)
RUN sed -i 's/getty 38400 tty1/agetty --autologin root tty1 linux/' /etc/inittab

# Adding networking.sh script (works only on lts kernel yet)
# (https://github.com/copy/v86/blob/master/tools/docker/debian/networking.sh)
RUN if [ "$KERNEL" == "lts" ]; then \ 
    echo -e "rmmod ne2k-pci && modprobe ne2k-pci\nhwclock -s\nsetup-interfaces -a -r" > /root/networking.sh && \ 
    chmod +x /root/networking.sh; \ 
fi

# Enable minimal needed OpenRC services
# (see https://wiki.alpinelinux.org/wiki/Alpine_Linux_in_a_chroot#Preparing_init_services)
RUN for i in devfs dmesg mdev hwdrivers; do rc-update add $i sysinit; done
RUN for i in hwclock modules sysctl hostname syslog bootmisc; do rc-update add $i boot; done
RUN rc-update add killprocs shutdown

# Generate initramfs with 9p modules
RUN mkinitfs -F "ata base ide scsi virtio ext4 9p" $(cat /usr/share/kernel/$KERNEL/kernel.release)
```

**2.** Create image container following by [tools/docker/debian/build-container.sh](https://github.com/copy/v86/blob/master/tools/docker/debian/build-container.sh)

https://github.com/copy/v86/blob/360d5691ff2ef0ca1c1a16d683910338b7325f07/tools/docker/debian/build-container.sh#L11-L16

```bash
"$(dirname "$0")"/../../../tools/fs2json.py --exclude /.dockerenv --out "$OUT_FSJSON" "$OUT_ROOTFS_TAR"
# make sure that .dockerenv not included in fs.json (if this happened, remove it from "fsroot" in json)
```

https://github.com/copy/v86/blob/360d5691ff2ef0ca1c1a16d683910338b7325f07/tools/docker/debian/build-container.sh#L20-L22

**3.** Extract kernel and initrd
```bash
tar -zxf "$OUT_ROOTFS_TAR" --strip-components=1 boot/*-<lts or virt>
```

**4.** Add to `V86` starter on your page (or use [examples/arch.html](https://github.com/copy/v86/blob/master/examples/arch.html)):
```javascript
  var emulator = new V86({
        ...
        bzimage: {
            url: "vmlinuz-...",
        },
        initrd: {
            url: "initramfs-...",
        },
        filesystem: {
            baseurl: "path/to/OUT_ROOTFS_FLAT/",
            basefs: "path/to/OUT_FSJSON",
        },
        bzimage_initrd_from_filesystem: false,
        cmdline: [
            "rw",
            "root=host9p rootfstype=9p rootflags=trans=virtio,cache=loose",
            "nowatchdog"
        ].join(" "),
    });
```

-----------------------------
Notes:
 - ~~Alpine Linux v3.19.x is not working, it has some problems with writing files into 9p root, that ruined start some services in OpenRC~~ use a https://github.com/copy/v86/commit/9286ca814a342e358e2ca05e4089e3bc35a03864 (you need build a v86 from `wip` branch)
 - Firefox has a long startup problem, probably by #797
 - For network, set `networking_proxy` url and run `/root/networking.sh`

**Reply by owenthereal** [reply]
@SuperMaxusa:

Can I clarify the values for these two fields? I couldn't find any references:

```
 bzimage: {
            url: "vmlinuz-...",
        },
        initrd: {
            url: "initramfs-...",
        },
```

And I followed [your guide](https://github.com/copy/v86/blob/master/tools/docker/alpine/Readme.md), but I couldn't get it to work. Would it be possible to provide the `build-state.js` script and add the `alpine.html` like the Debian one? I appreciate any help you can provide.

**Reply by SuperMaxusa** [reply]
> Can I clarify the values for these two fields? I couldn't find any references

On step 3 in first version need to extract from rootfs tar files from `/boot/`, and you have `vmlinuz-<virt or lts from $KERNEL>` (**bzimage** field) and `initramfs-<virt or lts>` (**initrd** field), next step - insert URL paths to these files in fiels, as step 4.

https://github.com/copy/v86/blob/2bcfa9d8c427f5b1d395b8ff931751550ce25e42/src/browser/starter.js#L31-L34

However v86 can automatically use kernel and initrd from 9p filesystem, so you can just leave `bzimage_initrd_from_filesystem: true`, as that mentioned in [alpine.md](https://github.com/copy/v86/blob/master/tools/docker/alpine/Readme.md).

https://github.com/copy/v86/blob/2bcfa9d8c427f5b1d395b8ff931751550ce25e42/tools/docker/alpine/Readme.md?plain=1#L8-L13

> And I followed your guide, but I couldn't get it to work.

Can you tell more about this please, it's about building a filesystem or running into browser/nodejs? I'm have tested now on new cloned repo and it's worked for me without any problems (make sure that your tar from coreutils, Busybox's tar can't delete files into archive):
```shell
$ git clone https://github.com/copy/v86.git
$ cd v86/tools/docker/alpine
$ chmod +x build-container.sh
$ ./build-container.sh
```

> Would it be possible to provide the build-state.js script and add the alpine.html like the Debian one?

I'm forget to add autologin on ttyS0, so build-state.js will not works. Here is branch in my fork that has this: ~~https://github.com/SuperMaxusa/v86/tree/alpine-docs~~ outdated, see #1046

**Reply by farhan-ct** [reply]
@SuperMaxusa is the 'Invalid argument' issue still there for you when the image is created with 9p?

**Reply by SuperMaxusa** [reply]
Yes, in Alpine Linux only branch `v3.18` has kernel version `6.1.86-r0` at this moment: https://pkgs.alpinelinux.org/package/v3.18/main/x86/linux-lts

**Reply by progrium** [reply]
when I change the Dockerfile in `wip` for alpine to use 3.18, it fails to mount root. Anybody know what that's about?
<img width="848" alt="Screenshot 2024-05-20 at 5 47 12 PM" src="https://github.com/copy/v86/assets/647/178c7d54-e885-43bf-8127-36fe83158a09">

**Reply by copy** [reply]
@progrium I can't reproduce this (although I had to use 3.18.3 since 3.18.1 wasn't downloading for me). Could you try:

- Show the output of `lsmod`
- `modprobe virtio virtio_pci; mkdir -p /mnt; mount -t 9p host9p /mnt`
- Delete the built files and rebuild from scratch

**Reply by SuperMaxusa** [reply]
> when I change the Dockerfile in `wip` for alpine to use 3.18, it fails to mount root

If you use a **3.18.x**, did you leave this line for installing mkinitfs from `edge` branch in Dockerfile?
https://github.com/copy/v86/blob/c69cbd7854d7cbda08679c389973aa87c9fb37bd/tools/docker/alpine/Dockerfile#L14-L15

The repository for `3.18` contains an old version of [mkinitfs](https://pkgs.alpinelinux.org/package/v3.18/main/x86/mkinitfs), that doesn't have a 9p support (https://github.com/alpinelinux/mkinitfs/commit/2a38df8c1db98b1bd9455b02086c77354d92574f, https://pkgs.alpinelinux.org/contents?file=9p.modules&name=mkinitfs&branch=v3.18).

**Reply by progrium** [reply]
Yes, it looks like I was slightly out of date, and using that Dockerfile with mkinitfs from edge fixed the problem. Thank you!
